### ISSUE TITLE
ec2_prefer_imdsv2 is set to true in configuration when it is not

### ISSUE THREAD

**Issue by prune998** [DataDog/datadog-agent#6523]
**Output of the info page (if this is a bug)**
```
(Paste the output of the info page here)
```

**Describe what happened:**
When starting an agent on an EC2 instance, it keeps reporting `ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token`.
This parameter is not set, and the error keeps happening even if I set env var `DD_EC2_PREFER_IMDSV2=false`

(issue also happen with a valid API key)

**Describe what you expected:**
I expect the datadog-agent to call the EC2 API using the old protocol (imdsv1) and not log an error. I would also expect the agent to get tags and other metadata from the EC2 API instead of `unable to get tags from aws`

**Steps to reproduce the issue:**

Start the docker datadog-agent on a Linux server: 
```
docker run  --rm   --volume /var/run/docker.sock:/var/run/docker.sock:ro   --volume /proc/:/host/proc/:ro   --volume /sys/fs/cgroup/:/host/sys/fs/cgroup:ro  -e DD_API_KEY=none  -e DD_LOG_LEVEL=warn  --name=datadog-agent   datadog/agent:latest
```

**Additional environment details (Operating System, Cloud provider, etc):**
Linux Centos 7 and Ubuntu.

Agent Version: `Agent 7.23.0 - Commit: 3cc240d - Serialization version: v4.44.0 - Go version: go1.14.7`

Full logs:

```
docker run  --rm  --volume /var/run/docker.sock:/var/run/docker.sock:ro   --volume /proc/:/host/proc/:ro   --volume /sys/fs/cgroup/:/host/sys/fs/cgroup:ro  -e DD_API_KEY=none  -e DD_LOG_LEVEL=warn  --name=datadog-agent   datadog/agent:latest
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-check-apikey.sh: executing...
[cont-init.d] 01-check-apikey.sh: exited 0.
[cont-init.d] 50-ci.sh: executing...
[cont-init.d] 50-ci.sh: exited 0.
[cont-init.d] 50-cri.sh: executing...
[cont-init.d] 50-cri.sh: exited 0.
[cont-init.d] 50-ecs.sh: executing...
[cont-init.d] 50-ecs.sh: exited 0.
[cont-init.d] 50-eks.sh: executing...
[cont-init.d] 50-eks.sh: exited 0.
[cont-init.d] 50-kubernetes.sh: executing...
[cont-init.d] 50-kubernetes.sh: exited 0.
[cont-init.d] 50-mesos.sh: executing...
[cont-init.d] 50-mesos.sh: exited 0.
[cont-init.d] 51-docker.sh: executing...
[cont-init.d] 51-docker.sh: exited 0.
[cont-init.d] 59-defaults.sh: executing...
[cont-init.d] 59-defaults.sh: exited 0.
[cont-init.d] 60-network-check.sh: executing...
[cont-init.d] 60-network-check.sh: exited 0.
[cont-init.d] 60-sysprobe-check.sh: executing...
grep: /etc/datadog-agent/system-probe.yaml: No such file or directory
grep: /etc/datadog-agent/system-probe.yaml: No such file or directory
[cont-init.d] 60-sysprobe-check.sh: exited 0.
[cont-init.d] 89-copy-customfiles.sh: executing...
[cont-init.d] 89-copy-customfiles.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
starting agent
starting security-agent
starting system-probe
starting trace-agent
starting process-agent
[services.d] done.
2020-10-08 15:21:10 UTC | PROCESS | WARN | (pkg/util/log/log.go:480 in func1) | failed to get configuration value for key "process_agent_config.host_ips": unable to cast <nil> of type <nil> to []string
2020-10-08 15:21:11 UTC | CORE | WARN | (pkg/forwarder/forwarder_health.go:209 in hasValidAPIKey) | api_key 'none' for domain https://api.datadoghq.com is invalid
2020-10-08 15:21:11 UTC | CORE | ERROR | (pkg/forwarder/forwarder_health.go:111 in healthCheckLoop) | No valid api key found, reporting the forwarder as unhealthy.
2020-10-08 15:21:12 UTC | CORE | WARN | (pkg/autodiscovery/providers/file.go:78 in Collect) | Skipping, open /opt/datadog-agent/bin/agent/dist/conf.d: no such file or directory
2020-10-08 15:21:12 UTC | CORE | WARN | (pkg/autodiscovery/providers/file.go:78 in Collect) | Skipping, open : no such file or directory
2020-10-08 15:21:13 UTC | CORE | WARN | (pkg/util/ec2/ec2.go:222 in doHTTPRequest) | ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token: ********
2020-10-08 15:21:14 UTC | CORE | WARN | (pkg/util/ec2/ec2.go:222 in doHTTPRequest) | ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token: ********
2020-10-08 15:21:14 UTC | CORE | WARN | (pkg/util/ec2/ec2.go:222 in doHTTPRequest) | ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token: ********
2020-10-08 15:21:14 UTC | CORE | WARN | (pkg/util/ec2/ec2_tags.go:90 in GetTags) | unable to get tags from aws and cache is empty: UnauthorizedOperation: You are not authorized to perform this operation.
        status code: 403, request id: 8f474bc5-5961-4f1c-8d5c-xxxx
2020-10-08 15:21:14 UTC | CORE | WARN | (pkg/util/gce/gce_tags.go:48 in getCachedTags) | unable to get tags from gce and cache is empty: status code 404 trying to GET http://169.254.169.254/computeMetadata/v1/?recursive=true
```

**Comment by olivielpeau** [comment]
Hi @prune998, thanks for reporting this issue.

We are currently investigating this and working on a fix. It seems to affect v`7.23.0`/ v`6.23.0` of the Agent. Until we release a fix, you should be able to use one of the following workarounds:

- pull the previous stable image of the Agent instead: `datadog/agent:7.22.1`, or
- modify the metadata of your EC2 instance so that it [allows additional hops](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsResponse.html#API_InstanceMetadataOptionsResponse_Contents) (typically, 2 hops). With the AWS CLI:
```
aws ec2 modify-instance-metadata-options  --instance-id <instance_id> --http-put-response-hop-limit 2
```

Once you've used one of these workarounds, please note that if you wish to collect EC2 instance tags from the Agent, you'll have to set the option `collect_ec2_tags: true` in the Agent's configuration (or the env var `DD_COLLECT_EC2_TAGS=true`).

Let me know how this works for you, thanks.

**Comment by prune998** [comment]
2nd option seems to be working fine.
Thanks.

**Comment by olivielpeau** [comment]
Thanks for confirming 👍 

Also, since my last message, we have determined that the warning `ec2_prefer_imdsv2 is set to true in configuration but the agent was unable to get a token` in `7.23.0` is misleading/wrong (and is indeed logged even when `ec2_prefer_imdsv2` is set to false/unset) and should not actually affect the collection of AWS instance tags from the Agent.

We'll fix this misleading warning in an upcoming `7.23.1` release, see related PR https://github.com/DataDog/datadog-agent/pull/6527.

**Comment by prune998** [comment]
That's what I thought after tcpdump-ing... :)
Still, the agent is pretty chatty for all sort of things that should even not be warnings (but infos). 

Whatever, close the issue when you think it's resolved (link to commit if possible). 
Thanks

**Comment by olivielpeau** [comment]
This misleading warning has been fixed via https://github.com/DataDog/datadog-agent/pull/6527, which has been released today in v7.23.1/v6.23.1.

Closing this, thanks again @prune998 for reporting.
